rfd/0228-resource-scoped-constraints-in-access-requests.md

```diff
@@ -0,0 +1,168 @@
+
+**User stories**:
+- **AWS Console (role elevation)**:  
+  Alice, a platform engineer with a standing read-only AWS role, needs elevated access to respond to an incident.  In Teleport's Web UI, she searches for the AWS Console app and clicks its connect dropdown. There, she sees both her read-only role (available right away) and the other roles she is able to request. She selects the 'admin' ARN, submits the request, and after approval, is able to assume the request and launch the console with that elevated ARN.
+- **AWS IAM IC (permission-set elevation)**:  
+  Ben operates in AWS IAM IC with a default `contributor` permission set but occasionally needs `BillingAdmin`. In Teleport's Web UI, he clicks the AWS IAM IC app's connect dropdown, selects `BillingAdmin` under the **requestable** section, submits the request, and once approved uses only that permission set without receiving extra IC permissions.
+- **SSH node (login + optional sudo)**:  
+  Diego can SSH to a service host as `deploy` but needs a temporary `admin` login (and possibly `sudo`) to perform maintenance. In Teleport's Web UI, he clicks the connect dropdown for the service node, and sees `deploy` under his **granted** roles, and `admin` (with an "allow sudo" checkbox) under the **requestable** section. He selects `admin`, checks "allow sudo", submits the request, and after approval, connects with just `admin` (and the approved sudo constraint).
+- **Database (role/user)**:  
+  Chandra has read-only access to a Postgres DB as `report_reader` and needs `migration_admin` for a one-off schema change. In Teleport's Web UI, she clicks the connect dropdown for the Postgres app, selects `migration_admin` under the **requestable** section, submits the request, and after approval, connects with only the `migration_admin` DB role for the session.
+
+**Reviewer experience**: Reviewers see both **(a)** the requested constraints and **b)** the resolved roles computed to satisfy them.
+
+## Design Overview
+
+### ResourceID & Constraints
+
+> Context: `ResourceID` identifies resources in Access Requests and is also used in certs (`AllowedResourceIDs`) to scope post-approval authorization.
+
+We add an optional, typed `Constraints` field to`ResourceID`. The content is domain-specific and enumerated.
+
+#### Go
+
+```go
+package types
+
+type ResourceID struct {
+  // 1..4 are existing fields.
+  Kind            string              `json:"kind"`
+  Name            string              `json:"name"`
+  ClusterName     string              `json:"clusterName"`
+  SubResourceName string              `json:"subResourceName,omitempty"`
+  // New: Optional constraints for this resource.
+  Constraints     *RequestConstraints `json:"constraints,omitempty"`
+}
+
+type ConstraintDomain int32
```

Comment:
I guess a potential concern we have to keep in mind is that encoding more stuff into certificates will increase the size of an X.509 extension which is limited. Is that why you chose int32 here instead of string?

Reply:
Yes, we've seen issues with large b64-encoded roles in certs leading to issues so trying to keep that in mind here w/r/t minimizing the amt of info we add.

Reply:
In this same vein, added a section about enforcing (what I think are) some sane max sizes for encoded RequestConstraints. LMK what you think there.